### DISCUSSION TITLE
Singularity Container broken?

### DISCUSSION THREAD

**Discussion by cstenkamp** [frankier/skelshop#31] [Q&A]
I really don't know much about Singularity, but when I try to run `singularity run --nv shub://frankier/skelshop:latest python -m skelshop --help` or any other skelshop-command, 
* When the container isn't downloaded yet, I get the errors
```
gzip: ./frankier-skelshop-master.img.gz: not in gzip format
ERROR: Could not locate downloaded image
ABORT: Aborting with RETVAL=255
```
* When it is, the errors are 
```
ERROR  : Failed to mount image in (read only): Invalid argument
ABORT  : Retval = 255
```

Is it something I'm doing wrong or something wrong with the container?

**Comment by frankier** [comment]
It works for me. What environment is this in? Can you run other Singularity images?

**Comment by cstenkamp** [comment]
It's on UCLA's Pisa. Singularity Version 2.3.1-dist. What version do you use to build?

**Comment by frankier** [comment]
It runs locally on my own machine with singularity version 3.6.4 and also on the Case Midwestern cluster (after running `module load singularity`). It's built on SingularityHub. That version of Singularity seems kind of old. You can also just run the images from DockerHub with Singularity. I'm not sure if it's actually a bit confusing to provide both.

**Comment by cstenkamp** [comment]
Huh, I didn't know about `module load singularity` - where does that come from?

**Comment by frankier** [comment]
Many clusters use this so you can load different versions of packages. Often you have to load cuda too. I don't know about UCLA though.

**Comment by frankier** [comment]
I think it basically just patches different software into your `$PATH`.

**Comment by cstenkamp** [comment]
Aight. God I feel so incredibly stupid trying to understand this..

Other thing - After all that is done, my latest problem when trying to run the singularity container on the Midwestern cluster is the filter command 
```
python -m skelshop filter --track --track-conf opt_lighttrack --pose-matcher-config work/gcn_config.yaml --shot-seg=psd --segs-file /home/cxs800/tiny_dataset/snek/tiny_ellen-Scenes.csv /home/cxs800/tiny_dataset/snek/tiny_ellen.untracked.h5 /home/cxs800/tiny_dataset/snek/tiny_ellen.opt_lighttrack.h5
```
failing with
```
Error executing rule skel_filter_csvshotseg_opt_lighttrack on cluster (jobid: 1, external: 15352426, jobscript: /home/cxs800/.snakemake/tmp.c12yy4yg/snakejob.skel_filter_csvshotseg_opt_lighttrack.1.sh). For error details see the cluster log and the log files of the involved rule(s).
```

```
rule skel_filter_csvshotseg_opt_lighttrack:
    input: work/gcn_config.yaml, /home/cxs800/tiny_dataset/snek/tiny_ellen.untracked.h5, /home/cxs800/tiny_dataset/snek/tiny_ellen-Scenes.csv
    output: /home/cxs800/tiny_dataset/snek/tiny_ellen.opt_lighttrack.h5
    jobid: 0
    wildcards: base=tiny_ellen

Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/skelshop/skelshop/__main__.py", line 2, in <module>
    from skelshop.cmd.skelshop import skelshop
  File "/opt/skelshop/skelshop/cmd/skelshop.py", line 9, in <module>
    from .calibrate import calibrate
  File "/opt/skelshop/skelshop/cmd/calibrate.py", line 22, in <module>
    from skelshop.face.pipe import (
  File "/opt/skelshop/skelshop/face/pipe.py", line 8, in <module>
    from face_recognition.api import (
  File "/usr/local/lib/python3.8/dist-packages/face_recognition/__init__.py", line 7, in <module>
    from .api import load_image_file, face_locations, batch_face_locations, face_landmarks, face_encodings, compare_faces, face_distance
  File "/usr/local/lib/python3.8/dist-packages/face_recognition/api.py", line 26, in <module>
    cnn_face_detector = dlib.cnn_face_detection_model_v1(cnn_face_detection_model)
RuntimeError: Error while calling cudaGetDevice(&the_device_id) in file /tmp/pip-req-build-gf1p5uv6/dlib/cuda/gpu_data.cpp:204. code: 35, reason: CUDA driver version is insufficient for CUDA runtime version
[Wed Nov 25 10:03:52 2020]
Error in rule skel_filter_csvshotseg_opt_lighttrack:
    jobid: 0
    output: /home/cxs800/tiny_dataset/snek/tiny_ellen.opt_lighttrack.h5
    shell:
        python -m skelshop filter --track --track-conf opt_lighttrack --pose-matcher-config work/gcn_config.yaml --shot-seg=psd --segs-file /home/cxs800/tiny_dataset/snek/tiny_ellen-Scenes.csv /home/cxs800/tiny_dataset/snek/tiny_ellen.untracked.h5 /home/cxs800/tiny_dataset/snek/tiny_ellen.opt_lighttrack.h5
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!
```

`CUDA driver version is insufficient for CUDA runtime version` - do you not run into this on the cluster?

**Comment by frankier** [comment]
No it is a bit fiddly at the moment.

First you need to load CUDA. Although that should be changed so face_detection is not imported unless used

`module load cuda/10.1`

**Comment by cstenkamp** [comment]
Ah, okay, thanks. Are there any other modules I need to load? :see_no_evil:

**Comment by frankier** [comment]
Not apart from `module load singularity` I don't think

**Comment by cstenkamp** [comment]
Hm, still the same error. Any thing else I might be overlooking?

**Comment by frankier** [comment]
You might need to use a GPU node. I've made it so that dlib is lazy imported now, and will tell you when the new Singularity container is built (still building on DockerHub atm) and this new container should hopefully let you do tracking on a CPU node.

It might be that we need multiple containers at some point for different configurations, but this is annoying since scuppers the current approach with Snakemake: https://frankier.github.io/skelshop/snakemake/ which relies on everything running in a single container and has worked nicely for me up until now. Snakemake supports orchestrating multiple Singularity containers but it definitely complicates things -- so that why I'm trying to stick with one container if at all possible. The idea currently is that any step that can use a GPU will require one, whereas other stuff should still be able to run on the CPU even though it's the "GPU container". We could try and switch between GPU/CPU dlib the same way as Peter's original OpenPose container did (and the multi OpenPose container based on this approach at https://github.com/frankier/openpose_containers ) but we potentially tie ourselves into having to build multiple versions of every library (not PyTorch since it autoconfigures itself for CPU/GPU) which also complicates things and is slow to test (not to mention that compiling dlib once takes long enough -- it increases container build times when building from scratch). Still trying to decide on the best way forward here. Open to input.

**Comment by frankier** [comment]
Built

**Comment by cstenkamp** [comment]
Nice, thanks!

But to select a GPU Node, I'd have to add that to your slurm-config, right?
So add it here, I suppose? https://github.com/frankier/skelshop/blob/master/contrib/slurm/skels.tracked.clusc.json#L17-L19

**Comment by frankier** [comment]
Yes but it shouldn't be needed now.

By the way, you don't need to rebuild everything to test changes to the `clusc.json` file. You can just change it on the host and point `CLUSC_CONF=$(pwd)/skels.tracked.clusc.json`. I'm not sure -- maybe the docs should recommend to copy that file into your working space as a first step rather than using the example one that just so happens to be inside the container.

**Comment by cstenkamp** [comment]
I know, I did that and just wanted to ensure that I did it correctly

**Comment by cstenkamp** [comment]
But yes, I only got that from the readme of your slurm-repo and mentioning it also where it's used would be helpful
